File: api/src/margin_api/schemas/model_validation.py
"""Model validation API response schemas."""

from __future__ import annotations

from pydantic import BaseModel


class MetricDistributionResponse(BaseModel):
    """Descriptive statistics for a single metric across seeds."""

    mean: float
    median: float
    std: float
    min: float
    max: float
    ci_lower: float
    ci_upper: float
    cv: float


class GateCheckResponse(BaseModel):
    """Result of a single promotion gate check."""

    name: str
    value: float
    threshold: float
    passed: bool


class ModelComparisonResponse(BaseModel):
    """Statistical comparison between candidate and incumbent model."""

    p_value: float
    effect_size: float
    significant: bool
    label: str
    n_compared: int
    mean_difference: float


class SeedDetailResponse(BaseModel):
    """Per-seed training result."""

    seed: int
    rank_ic: float
    n_clusters: int
    n_samples: int
    selected: bool


class SeedValidationReportResponse(BaseModel):
    """Full seed validation report for a single training run group."""

    run_group_id: str
    created_at: str
    n_seeds: int
    gate_passed: bool
    selected_seed: int | None
    metric_distributions: dict[str, MetricDistributionResponse]
    gate_checks: list[GateCheckResponse]
    seed_details: list[SeedDetailResponse]
    environment_snapshot: dict
    comparison: ModelComparisonResponse | None = None


class SeedValidationHistoryResponse(BaseModel):
    """Paginated list of seed validation reports."""

    reports: list[SeedValidationReportResponse]
    total: int

File: web/src/lib/api/model-validation.ts
import { apiFetch } from "./client"

export interface MetricDistribution {
  mean: number
  median: number
  std: number
  min: number
  max: number
  ci_lower: number
  ci_upper: number
  cv: number
}

export interface GateCheck {
  name: string
  value: number
  threshold: number
  passed: boolean
}

export interface ModelComparison {
  p_value: number
  effect_size: number
  significant: boolean
  label: string
  n_compared: number
  mean_difference: number
}

export interface SeedDetail {
  seed: number
  rank_ic: number
  n_clusters: number
  n_samples: number
  selected: boolean
}

export interface SeedValidationReport {
  run_group_id: string
  created_at: string
  n_seeds: number
  gate_passed: boolean
  selected_seed: number | null
  metric_distributions: Record<string, MetricDistribution>
  gate_checks: GateCheck[]
  seed_details: SeedDetail[]
  environment_snapshot: Record<string, unknown>
  comparison: ModelComparison | null
}

export interface ValidationHistory {
  reports: SeedValidationReport[]
  total: number
}

export async function getLatestValidationReport(): Promise<SeedValidationReport> {
  return apiFetch<SeedValidationReport>("/api/v1/admin/model-validation/latest")
}

export async function getValidationHistory(
  limit = 20,
  offset = 0,
): Promise<ValidationHistory> {
  return apiFetch<ValidationHistory>(
    `/api/v1/admin/model-validation/history?limit=${limit}&offset=${offset}`,
  )
}

export async function getValidationReport(
  runGroupId: string,
): Promise<SeedValidationReport> {
  return apiFetch<SeedValidationReport>(
    `/api/v1/admin/model-validation/${runGroupId}`,
  )
}
